README.rst
===========
coordinates
===========

*****
Usage
*****

A short usage example::

    from datetime import datetime
    from coordinates import satellite_xyz

    filename = 'brdm2510.17p'
    epoch = datetime(2017, 9, 8, 1, 50, 0)

    xyz = satellite_xyz(filename, 'G', 1, epoch)

************
Installation
************

Install package from GitHub in "editable" mode.

    $ pip install -e git+https://github.com/gnss-lab/coordinates.git#egg=coordinates

*******
License
*******

Distributed under the terms of the
`MIT <https://github.com/gnss-lab/gnss-tec/blob/master/LICENSE.txt>`_
license, gnss-tec is free and open source software.

Copyright Ilya Zhivetiev, 2017.

CHANGELOG.rst
coordinates v1.1.0
====================

- ``CoordinatesException`` -- the base class for other
  exceptions of the module.

coordinates v1.0.1
==================

- PEP 479 adaption.
- Add v2.01 to supported versions.


tests/testlib.py
from contextlib import contextmanager
from os import remove
from tempfile import NamedTemporaryFile


@contextmanager
def mktmp(tmp_str):
    tmp_file_name = None

    try:
        with NamedTemporaryFile(mode='w', delete=False) as tmp_file:
            tmp_file.writelines(tmp_str)
            tmp_file_name = tmp_file.name

        yield tmp_file_name

    finally:
        if tmp_file_name:
            remove(tmp_file_name)


tests/conftest.py
from io import StringIO

import pytest
from testlib import mktmp

_nav_content_v2 = '''\
     2              NAVIGATION DATA                         RINEX VERSION / TYPE
CCRINEXN V1.6.0 UX  CDDIS               12-APR-16 17:31     PGM / RUN BY / DATE 
IGS BROADCAST EPHEMERIS FILE                                COMMENT             
    0.1583D-07  0.1490D-07 -0.1192D-06 -0.1192D-06          ION ALPHA           
    0.1065D+06  0.6554D+05 -0.1966D+06 -0.1966D+06          ION BETA            
    0.186264514923D-08 0.000000000000D+00   233472     1892 DELTA-UTC: A0,A1,T,W
    17                                                      LEAP SECONDS        
                                                            END OF HEADER       
 1 16  4 11  0  0  0.0 0.169607810676D-04 0.113686837722D-11 0.000000000000D+00
    0.350000000000D+02 0.222500000000D+02 0.482198656938D-08 0.368417754673D+00
    0.121630728245D-05 0.527631223667D-02 0.668689608574D-05 0.515364252472D+04
    0.864000000000D+05 0.391155481338D-07-0.140616900879D+01 0.106170773506D-06
    0.963711739811D+00 0.247687500000D+03 0.450110393247D+00-0.827070165078D-08
    0.284654714154D-09 0.100000000000D+01 0.189200000000D+04 0.000000000000D+00
    0.200000000000D+01 0.000000000000D+00 0.512227416039D-08 0.350000000000D+02
    0.805020000000D+05 0.400000000000D+01 0.000000000000D+00 0.000000000000D+00
 2 16  4 11  0  0  0.0 0.597649253905D-03-0.181898940355D-11 0.000000000000D+00
    0.830000000000D+02 0.185937500000D+02 0.527057686384D-08 0.763017673126D+00
    0.114180147648D-05 0.156129685929D-01 0.783614814282D-05 0.515374914742D+04
    0.864000000000D+05-0.163912773132D-06-0.145069785349D+01 0.100582838059D-06
    0.942463959213D+00 0.223656250000D+03-0.213318032835D+01-0.861071569602D-08
    0.216080431326D-09 0.100000000000D+01 0.189200000000D+04 0.000000000000D+00
    0.200000000000D+01 0.000000000000D+00-0.200234353542D-07 0.830000000000D+02
    0.864000000000D+05 0.000000000000D+00 0.000000000000D+00 0.000000000000D+00
'''

_nav_content_v3 = '''\
     3.03           NAVIGATION DATA     M (Mixed)           RINEX VERSION / TYPE
BCEmerge            congo               20170909 012902 GMT PGM / RUN BY / DATE 
Merged GPS/GLO/GAL/BDS/QZS/SBAS/IRNSS navigation file       COMMENT 
based on CONGO and MGEX tracking data                       COMMENT 
DLR: O. Montenbruck;  TUM: P. Steigenberger                 COMMENT 
GAUT  2.7939677238e-09 1.776356839e-15 345600 1965          TIME SYSTEM CORR    
GLGP  1.8626451492e-08 0.000000000e+00 345600 1965          TIME SYSTEM CORR    
GLUT  9.3132257462e-10 0.000000000e+00 345600 1965          TIME SYSTEM CORR    
GPGA  1.0040821508e-08 3.330669074e-14 432000 1965          TIME SYSTEM CORR    
GPUT  9.3132257462e-10 1.776356839e-15 589824 1965          TIME SYSTEM CORR    
QZUT  6.5192580223e-09 0.000000000e+00   8192 1966          TIME SYSTEM CORR    
    18                                                      LEAP SECONDS        
                                                            END OF HEADER       
G01 2017 09 08 00 00 00 5.530333146453e-05-4.547473508865e-13 0.000000000000e+00
     7.200000000000e+01-2.653125000000e+01 4.883774857397e-09-2.309091328065e-01
    -1.233071088791e-06 7.021094555967e-03 2.788379788399e-06 5.153673307419e+03
     4.320000000000e+05 7.823109626770e-08 2.011192878165e+00 4.284083843231e-08
     9.681868691207e-01 3.310000000000e+02 6.192489497701e-01-8.576071513516e-09
    -2.500104139373e-11 1.000000000000e+00 1.965000000000e+03 0.000000000000e+00
     2.000000000000e+00 0.000000000000e+00 5.587935447693e-09 7.200000000000e+01
     4.248180000000e+05 4.000000000000e+00                                      
S20 2017 09 08 00 16 32 0.000000000000e+00 0.000000000000e+00 4.330030000000e+05
     4.063672000000e+04 0.000000000000e+00 0.000000000000e+00 1.000000000000e+00
    -1.124591600000e+04 0.000000000000e+00 0.000000000000e+00 3.276700000000e+04
     0.000000000000e+00 0.000000000000e+00 0.000000000000e+00 2.150000000000e+02
'''

_nav_content_v3_unsorted = '''\
     3.03           NAVIGATION DATA     M (Mixed)           RINEX VERSION / TYPE
BCEmerge            congo               20170909 012902 GMT PGM / RUN BY / DATE 
    18                                                      LEAP SECONDS        
                                                            END OF HEADER       
S20 2017 09 08 00 05 52 0.000000000000e+00 0.000000000000e+00 4.323620000000e+05
     4.063672000000e+04 0.000000000000e+00 0.000000000000e+00 1.000000000000e+00
    -1.124591600000e+04 0.000000000000e+00 0.000000000000e+00 3.276700000000e+04
     0.000000000000e+00 0.000000000000e+00 0.000000000000e+00 2.200000000000e+01
S20 2017 09 08 00 00 32 0.000000000000e+00 0.000000000000e+00 4.320410000000e+05
     4.063672000000e+04 0.000000000000e+00 0.000000000000e+00 1.000000000000e+00
    -1.124591600000e+04 0.000000000000e+00 0.000000000000e+00 3.276700000000e+04
     0.000000000000e+00 0.000000000000e+00 0.000000000000e+00 2.000000000000e+00
S20 2017 09 08 00 03 12 0.000000000000e+00 0.000000000000e+00 4.322000000000e+05
     4.063672000000e+04 0.000000000000e+00 0.000000000000e+00 1.000000000000e+00
    -1.124591600000e+04 0.000000000000e+00 0.000000000000e+00 3.276700000000e+04
     0.000000000000e+00 0.000000000000e+00 0.000000000000e+00 1.200000000000e+01
'''


@pytest.fixture
def nav_file_v2():
    return mktmp(_nav_content_v2)


@pytest.fixture
def nav_iter_v2():
    return StringIO(_nav_content_v2)


@pytest.fixture
def nav_file_v3():
    return mktmp(_nav_content_v3)


@pytest.fixture
def nav_iter_v3():
    return StringIO(_nav_content_v3)


@pytest.fixture
def nav_file_unsorted_v3():
    return mktmp(_nav_content_v3_unsorted)


tests/test_sat.py
import datetime

import pytest

from coordinates.exceptions import SatSystemError, NavMessageNotFoundError
from coordinates.sat import GLO_WAY, GPS_WAY
from coordinates.sat import (
    find_message,
    get_day_sec,
    get_dt,
    get_week_sec,
    glo_sat_xyz,
    gps_sat_xyz,
    nearest_message,
    read_nav_data,
    reference_time,
    xyz_calculator,
)


@pytest.fixture
def unknown_sat_system():
    return 'X'


def test_get_week_sec():
    epoch_start = datetime.datetime(1980, 1, 6, 0, 0, 0)
    epoch = datetime.datetime(1980, 1, 8, 0, 0, 30)

    std = 172830  # 2 days and 30 seconds
    test = get_week_sec(epoch, epoch_start)

    assert std == test


def test_get_day_sec():
    epoch = datetime.datetime(1980, 1, 6, 0, 2, 30)

    std = 150
    test = get_day_sec(epoch)

    assert std == test


def test_reference_time(unknown_sat_system):
    with pytest.raises(SatSystemError, match=unknown_sat_system):
        reference_time(unknown_sat_system, datetime.datetime.now())


def test_read_nav_data(nav_file_unsorted_v3):
    std = {
        ('S', 20): [
            {
                'epoch': datetime.datetime(2017, 9, 8, 0, 0, 32),
                'message': (40636.72, 0., 0., 1.0,
                            -11245.916, 0., 0., 32767.0,
                            0., 0., 0., 2.0)
            },
            {
                'epoch': datetime.datetime(2017, 9, 8, 0, 3, 12),
                'message': (40636.72, 0., 0., 1.0,
                            -11245.916, 0., 0., 32767.0,
                            0., 0., 0., 12.)
            },
            {
                'epoch': datetime.datetime(2017, 9, 8, 0, 5, 52),
                'message': (40636.72, 0., 0., 1.0,
                            -11245.916, 0., 0., 32767.0,
                            0., 0., 0., 22.)
            },
        ],

    }

    with nav_file_unsorted_v3 as filename:
        test = read_nav_data(filename)

    assert std == test


def test_nearest_message():
    messages = [
        {'epoch': datetime.datetime(2017, 9, 8, 2, 0), 'message': 1},
        {'epoch': datetime.datetime(2017, 9, 8, 4, 0), 'message': None},
        {'epoch': datetime.datetime(2017, 9, 8, 6, 0), 'message': None},
        {'epoch': datetime.datetime(2017, 9, 8, 8, 0), 'message': None},
        {'epoch': datetime.datetime(2017, 9, 8, 10, 0), 'message': None},
        {'epoch': datetime.datetime(2017, 9, 8, 12, 0), 'message': None},
        {'epoch': datetime.datetime(2017, 9, 8, 14, 0), 'message': None},
        {'epoch': datetime.datetime(2017, 9, 8, 15, 59, 44), 'message': 3},
        {'epoch': datetime.datetime(2017, 9, 8, 16, 0), 'message': None},
        {'epoch': datetime.datetime(2017, 9, 8, 17, 59, 44), 'message': None},
        {'epoch': datetime.datetime(2017, 9, 8, 20, 0), 'message': None},
        {'epoch': datetime.datetime(2017, 9, 8, 22, 0), 'message': 2},
    ]

    epoch = datetime.datetime(2017, 9, 8, 2, 15)
    assert nearest_message(messages, epoch) == messages[0]

    epoch = datetime.datetime(2017, 9, 8, 0, 0)
    assert nearest_message(messages, epoch) == messages[0]

    epoch = datetime.datetime(2017, 9, 8, 23, 55)
    assert nearest_message(messages, epoch) == messages[-1]

    epoch = datetime.datetime(2017, 9, 8, 15, 59, 44)
    assert nearest_message(messages, epoch) == messages[7]

    with pytest.raises(NavMessageNotFoundError):
        epoch = datetime.datetime(2017, 9, 7)
        nearest_message(messages, epoch)


def test_get_dt():
    obs_epoch = datetime.datetime(2017, 1, 1, 0, 0, 0)
    msg_epoch = datetime.datetime(2017, 1, 1, 0, 0, 30)

    std = -30
    test = get_dt(msg_epoch, obs_epoch)
    assert std == test

    obs_epoch = datetime.datetime(2017, 1, 1, 0, 5, 30)
    msg_epoch = datetime.datetime(2017, 1, 1, 0, 0, 0)

    std = 330
    test = get_dt(msg_epoch, obs_epoch)
    assert std == test


def test_find_message(unknown_sat_system):
    nav_data = {
        ('S', 38): [
            {'epoch': datetime.datetime(2017, 9, 8, 0, 1, 4), 'message': ()},
            {'epoch': datetime.datetime(2017, 9, 8, 0, 5, 20), 'message': ()},
            {'epoch': datetime.datetime(2017, 9, 8, 0, 9, 36), 'message': ()},
            {'epoch': datetime.datetime(2017, 9, 8, 0, 13, 52), 'message': ()},
            {'epoch': datetime.datetime(2017, 9, 8, 0, 18, 8), 'message': ()},
            {'epoch': datetime.datetime(2017, 9, 8, 0, 22, 24), 'message': ()},
            {'epoch': datetime.datetime(2017, 9, 8, 0, 26, 40), 'message': ()},
            {'epoch': datetime.datetime(2017, 9, 8, 0, 30, 56), 'message': ()},
            {'epoch': datetime.datetime(2017, 9, 8, 0, 35, 12), 'message': ()},
            {'epoch': datetime.datetime(2017, 9, 8, 0, 39, 28),
             'message': ('SBAS',)},
            {'epoch': datetime.datetime(2017, 9, 8, 0, 43, 44), 'message': ()},
            {'epoch': datetime.datetime(2017, 9, 8, 0, 48), 'message': ()},
        ],
        ('G', 23): [
            {'epoch': datetime.datetime(2017, 9, 8, 2, 0),
             'message': ('GPS',)},
            {'epoch': datetime.datetime(2017, 9, 8, 4, 0), 'message': ()},
            {'epoch': datetime.datetime(2017, 9, 8, 6, 0), 'message': ()},
            {'epoch': datetime.datetime(2017, 9, 8, 8, 0), 'message': ()},
            {'epoch': datetime.datetime(2017, 9, 8, 10, 0), 'message': ()},
            {'epoch': datetime.datetime(2017, 9, 8, 11, 59, 44),
             'message': ()},
            {'epoch': datetime.datetime(2017, 9, 8, 12, 0), 'message': ()},
            {'epoch': datetime.datetime(2017, 9, 8, 13, 59, 44),
             'message': ()},
            {'epoch': datetime.datetime(2017, 9, 8, 16, 0), 'message': ()},
            {'epoch': datetime.datetime(2017, 9, 8, 18, 0), 'message': ()},
            {'epoch': datetime.datetime(2017, 9, 8, 20, 0), 'message': ()},
            {'epoch': datetime.datetime(2017, 9, 8, 22, 0), 'message': ()},
        ],
    }

    # GLO-way
    satellite, number = 'S', 38
    epoch = datetime.datetime(2017, 9, 8, 0, 40)

    std = (32, ('SBAS',))
    test = find_message(nav_data, satellite, number, epoch)

    assert std == test

    # GPS-way
    satellite, number = 'G', 23

    std = (434400.0000000393, ('GPS',))
    test = find_message(nav_data, satellite, number, epoch)

    assert std == test

    with pytest.raises(NavMessageNotFoundError):
        find_message(nav_data, 'R', 1, epoch)

    with pytest.raises(SatSystemError, match=unknown_sat_system):
        find_message(nav_data, unknown_sat_system, 1, epoch)


def test_xyz_calculator(unknown_sat_system):
    for s in GPS_WAY:
        assert xyz_calculator(s) is gps_sat_xyz

    for s in GLO_WAY:
        assert xyz_calculator(s) is glo_sat_xyz

    with pytest.raises(SatSystemError, match=unknown_sat_system):
        xyz_calculator(unknown_sat_system)


coordinates/__init__.py
# coding=utf-8
"""
Tools for manipulating coordinates for gnss-lab project.

"""
from math import pi, sin, cos, atan2, sqrt

from coordinates.exceptions import XYZNotFoundError
from coordinates.sat import satellite_xyz

__all__ = ['satellite_xyz', 'retrieve_xyz', 'xyz2lbh']

__version__ = '1.1.0b1'
__author__ = __maintainer__ = 'Ilya Zhivetiev'
__email__ = 'i.zhivetiev@gnss-lab.org'


def retrieve_xyz(rnx_file):
    """Retrieve and return geocentric approximate marker position from the
    rnx_file.

    Parameters
    ----------
    rnx_file : str or file
        File or filename. The file must be a plain RINEX
        observation file, e.g. guzniped and crx2rnx-ed. Note
        that the position will seek to the start of the file.

    Returns
    -------
    xyz : tuple
        X, Y, and Z values

    Raises
    ------
    XYZNotFoundError
        'APPROX POSITION XYZ' record not found.
    """
    xyz_label = 'APPROX POSITION XYZ'
    end_label = 'END OF HEADER'

    xyz = None
    is_string = True

    try:
        rnx_file + ''
    except (TypeError, ValueError):
        is_string = False

    if is_string:
        rnx = open(rnx_file)
    else:
        rnx = iter(rnx_file)

    for line in rnx:
        label = line[60:].rstrip().upper()
        if label == xyz_label:
            xyz = [float(line[i:i + 14]) for i in range(0, 42, 14)]
            break
        if label == end_label:
            break

    if is_string:
        rnx.close()
    else:
        rnx.seek(0)

    if xyz:
        return tuple(xyz)
    else:
        raise XYZNotFoundError("Can't find %s" % xyz_label)


def xyz2lbh(x, y, z, deg=True):
    """Converts cartesian coordinates to geodetic coordinates.

    Parameters
    ----------
    x, y, z : float
        meters

    deg : bool, optional
        If True, l and b values will be converted into grad. Default is True.

    Returns
    -------
    l, b : float
        longitude and latitude

    h : float
        height, meters

    Note
    ----
    К.Ф. Афонин Высшая геодезия. Системы координат и преобразования
        между ними // Новосибирск. СГГА. 2011 г. 55 с.
    """

    # semiaxis and eccentricity of the ellipse
    datum_a = 6378137.00
    datum_b = 6356752.314245
    datum_e = sqrt(1 - pow(datum_b, 2) / pow(datum_a, 2))

    # threshold
    e_B = 1e-12

    Q = sqrt(x ** 2 + y ** 2)

    # L - longitude
    if x == 0:
        if y > 0:
            L = pi / 2
        else:
            L = 3 * pi / 2
    else:
        L = atan2(y, x)

    # B - latitude
    # initial values
    Bi1 = z / Q * 1 / (1 - datum_e ** 2)
    Bi = Bi1
    N = 0

    while 1:
        W = sqrt(1 - datum_e ** 2 * sin(Bi) ** 2)
        N = datum_a / W
        T = z + N * datum_e ** 2 * sin(Bi)

        Bi1 = atan2(T, Q)

        if abs(Bi1 - Bi) <= e_B:
            break

        Bi = Bi1

    B = Bi1

    # H - height
    H = Q * cos(B) + z * sin(B) - N * (1 - datum_e ** 2 * sin(B) ** 2)

    if deg:
        L, B = L * 180 / pi, B * 180 / pi

        if L < 0:
            L += 360

    return L, B, H


coordinates/sat.py
import datetime
from collections import defaultdict
from functools import lru_cache
from math import sqrt, sin, cos, atan2
from operator import itemgetter

from coordinates import datum
from coordinates.broadcast import rnx_nav
from coordinates.exceptions import SatSystemError, NavMessageNotFoundError

# GPS, BDS, Galileo, and IRNSS
GPS_WAY = {'G', 'C', 'E', 'I'}

# GLONASS, SBAS, and QZSS
GLO_WAY = {'R', 'S', 'J'}

KNOWN_SYSTEMS = GPS_WAY | GLO_WAY

EPOCH_START = dict(
    G=datetime.datetime(1980, 1, 6, 0, 0, 0),  # GPS
    C=datetime.datetime(2006, 1, 1, 0, 0, 0),  # BDS
    E=datetime.datetime(1999, 8, 22, 0, 0, 13),  # Galileo
    I=datetime.datetime(1999, 8, 21, 23, 59, 47),  # IRNSS
)


def get_week_sec(epoch, epoch_start):
    """Returns seconds since the beginning of the week

    Parameters
    ----------
    epoch : datetime.datetime
    epoch_start : datetime.datetime

    Returns
    -------
    seconds : float

    """
    current_epoch = epoch - epoch_start
    week = current_epoch.days / 7.

    seconds = (week - int(week)) * (7 * 24 * 60 * 60)
    seconds += current_epoch.seconds

    return seconds


def get_day_sec(epoch):
    """Returns seconds since the beginning of the day

    """
    time = epoch.time()
    seconds = datetime.timedelta(0, (time.hour * 60 ** 2 +
                                     time.minute * 60 +
                                     time.second)).seconds
    return seconds


def get_dt(msg_epoch, obs_epoch):
    """Returns delta between msg_epoch and obs_epoch in seconds.

    """
    if obs_epoch >= msg_epoch:
        return (obs_epoch - msg_epoch).seconds
    else:
        return (msg_epoch - obs_epoch).seconds * -1


def reference_time(satellite, epoch):
    """Returns reference time of the epoch according to satellite system in
    seconds.

    """
    if satellite not in EPOCH_START:
        raise SatSystemError(satellite)
    return get_week_sec(epoch, EPOCH_START[satellite])


def nearest_message(messages, obs_epoch):
    """Returns the nearest navigation message.

    """
    first_date = messages[0]['epoch'].date()
    if not obs_epoch.date() == first_date:
        raise NavMessageNotFoundError(
            'The dates of the nav message and observation must be the same.'
        )

    left = -1
    right = len(messages)

    while right - left > 1:
        pivot = (left + right) // 2
        if obs_epoch >= messages[pivot]['epoch']:
            left = pivot
        else:
            right = pivot

    if left == -1:
        return messages[0]

    return messages[left]


def find_message(nav_data, satellite, number, epoch):
    """Returns a navigation message and timedelta between the message
    time and the epoch.

    """
    if satellite not in KNOWN_SYSTEMS:
        raise SatSystemError(satellite)

    try:
        satellite_messages = nav_data[(satellite, number)]
    except KeyError:
        msg = 'No such satellite: {sat}{num}'.format(
            sat=satellite,
            num=number,
        )
        raise NavMessageNotFoundError(msg)

    # Returns the first for GPS, BDS, Galileo, and IRNSS
    if satellite in GPS_WAY:
        message = satellite_messages[0]
        # время с начала недели
        dt = reference_time(satellite, epoch)

    # Returns the nearest for GLONASS, SBAS, and QZSS
    elif satellite in GLO_WAY:
        message = nearest_message(satellite_messages, epoch)
        # разница между сообщением и набл
        dt = get_dt(message['epoch'], epoch)
    else:
        raise NavMessageNotFoundError(
            'Navigation message not found: {sat}{num} {epoch}'.format(
                sat=satellite,
                num=number,
                epoch=epoch,
            )
        )

    return dt, message['message']


def gps_sat_xyz(ephemeris, sec):
    """Returns geocentric coordinates XYZ of the GPS (GPS-way) satellite

    Parameters
    ----------
    ephemeris : list
        list with ephemeris

    sec : float
        amount of seconds since the start of the week, seconds

    Returns
    -------
    x : float
        X, meters
    y : float
        Y, meters
    z : float
        Z, meters

    """
    # threshold
    # e_ps = 2e-15

    crs = ephemeris[1]
    dn = ephemeris[2]
    m0 = ephemeris[3]
    cuc = ephemeris[4]
    e0 = ephemeris[5]
    cus = ephemeris[6]
    a0 = ephemeris[7] ** 2
    toe = ephemeris[8]
    cic = ephemeris[9]
    omega_0 = ephemeris[10]
    cis = ephemeris[11]
    i0 = ephemeris[12]
    crc = ephemeris[13]
    w0 = ephemeris[14]
    omega_dot = ephemeris[15]
    i_dot = ephemeris[16]

    tk = sec - toe

    if tk > 302400:
        tk -= 604800
    elif tk < -302400:
        tk += 604800

    n = sqrt(datum.mu / a0 ** 3) + dn
    mk = m0 + n * tk

    ek = mk
    ek_1 = ek + 1

    prv_d_ek = 0
    cur_d_ek = 1
    #   while ( abs(ek-ek_1) > e_ps ):
    while prv_d_ek != cur_d_ek:
        prv_d_ek = abs(ek - ek_1)
        ek_1 = ek
        ek = ek_1 - (ek_1 - e0 * sin(ek_1) - mk) / (1 - e0 * cos(ek_1))
        cur_d_ek = abs(ek - ek_1)
        # print prv_d_ek, cur_d_ek

    fs = (sqrt(1 - e0 ** 2) * sin(ek)) / (1 - e0 * cos(ek))
    fc = (cos(ek) - e0) / (1 - e0 * cos(ek))
    tettak = atan2(fs, fc)

    u0k = tettak + w0
    uk = u0k + cuc * cos(2 * u0k) + cus * sin(2 * u0k)

    rk = a0 * (1 + e0 * cos(ek)) + crc * cos(2 * u0k) + crs * sin(2 * u0k)

    ik = i0 + (cic * cos(2 * u0k) + cis * sin(2 * u0k)) + i_dot * tk

    omega_k = omega_0 + (omega_dot - datum.omega) * tk - datum.omega * toe

    x = rk * (cos(uk) * cos(omega_k) - sin(uk) * sin(omega_k) * cos(ik))
    y = rk * (cos(uk) * sin(omega_k) + sin(uk) * cos(omega_k) * cos(ik))
    z = rk * sin(uk) * sin(ik)

    return x, y, z


def glo_sat_xyz(ephemeris, dt):
    """Returns geocentric coordinates XYZ of the GLONASS (GLO-way) satellite

    Parameters
    ----------
    ephemeris : list
        list with ephemeris

    dt : float
        difference between time of the ephemeris and observation time, seconds

    Returns
    -------
    x : float
        X, meters
    y : float
        Y, meters
    z : float
        Z, meters
    """

    # meters
    x0 = ephemeris[0] * 1000
    vX = ephemeris[1] * 1000
    aX = ephemeris[2] * 1000

    y0 = ephemeris[4] * 1000
    vY = ephemeris[5] * 1000
    aY = ephemeris[6] * 1000

    z0 = ephemeris[8] * 1000
    vZ = ephemeris[9] * 1000
    aZ = ephemeris[10] * 1000

    if dt == 0:
        return x0, y0, z0

    r = sqrt(x0 ** 2 + y0 ** 2 + z0 ** 2)

    # - GLONASS ICD ver 5.1, 2008.
    first_sd = -(datum.mu / r ** 3)
    second_sd = 3 / 2. * datum.J0sqd * datum.mu * datum.ae ** 2 / r ** 5

    # - x, y, z
    fx = lambda x, z: (first_sd * x - second_sd * x *
                       (1 - 5 * z ** 2 / r ** 2) +
                       datum.omega ** 2 * x + 2 * datum.omega * vY + aX)

    fy = lambda y, z: (first_sd * y - second_sd * y *
                       (1 - 5 * z ** 2 / r ** 2) +
                       datum.omega ** 2 * y - 2 * datum.omega * vX + aY)

    fz = lambda z: (first_sd * z - second_sd * z *
                    (3 - 5 * z ** 2 / r ** 2) + aZ)

    x = dict()
    kx = dict()

    y = dict()
    ky = dict()

    z = dict()
    kz = dict()

    # 1
    kx[1] = fx(x0, z0) * dt
    ky[1] = fy(y0, z0) * dt
    kz[1] = fz(z0) * dt

    x[1] = x0 + vX * dt / 2. + kx[1] * dt / 8.
    y[1] = y0 + vY * dt / 2. + ky[1] * dt / 8.
    z[1] = z0 + vZ * dt / 2. + kz[1] * dt / 8.

    # 2
    kx[2] = fx(x[1], z[1]) * dt
    ky[2] = fy(y[1], z[1]) * dt
    kz[2] = fz(z[1]) * dt

    x[2] = x0 + vX * dt / 2. + kx[2] * dt / 8.
    y[2] = y0 + vY * dt / 2. + ky[2] * dt / 8.
    z[2] = z0 + vZ * dt / 2. + kz[2] * dt / 8.

    # 3
    kx[3] = fx(x[2], z[2]) * dt
    ky[3] = fy(y[2], z[2]) * dt
    kz[3] = fz(z[2]) * dt

    # result
    xyz = (
        x0 + vX * dt + (kx[1] + kx[2] + kx[3]) * dt / 6,
        y0 + vY * dt + (ky[1] + ky[2] + ky[3]) * dt / 6,
        z0 + vZ * dt + (kz[1] + kz[2] + kz[3]) * dt / 6
    )

    return xyz


def xyz_calculator(satellite):
    """Возвращает калькулятор XYZ в зависимости от спутниковой системы.

    """
    if satellite in GPS_WAY:
        return gps_sat_xyz
    elif satellite in GLO_WAY:
        return glo_sat_xyz
    else:
        raise SatSystemError(satellite)


@lru_cache(maxsize=8)
def read_nav_data(filename):
    """Returns dictionary which contains navigation data from the file.
    Navigation records are sorted by epoch.

    """
    nav_data = defaultdict(list)
    for row in rnx_nav(filename):
        satellite, number, epoch, sv_clock, message = row
        record = {'epoch': epoch, 'message': message}
        nav_data[(satellite, number)].append(record)

    for sat in nav_data:
        nav_data[sat].sort(key=itemgetter('epoch'))

    return nav_data


@lru_cache(maxsize=None)
def satellite_xyz(filename, satellite, number, epoch):
    """Returns XYZ coordinates of the satellite with number

    """
    calculate = xyz_calculator(satellite)
    data = read_nav_data(filename)
    dt, message = find_message(
        data,
        satellite,
        number,
        epoch,
    )
    xyz = calculate(message, dt)
    return xyz
